### PR TITLE
compression: add support for estargz

### PR DESCRIPTION
**Pull request by giuseppe**
add support for the estargz compression format from
https://github.com/containerd/stargz-snapshotter

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: pkg/compression/compression.go [hunk 1]
```diff
@@ -35,13 +37,17 @@ var (
 	// Zstd:chunked compression.
 	ZstdChunked = internal.NewAlgorithm(types.ZstdChunkedAlgorithmName, types.ZstdAlgorithmName, /* Note: InternalUnstableUndocumentedMIMEQuestionMark is not ZstdChunkedAlgorithmName */
 		nil, ZstdDecompressor, compressor.ZstdCompressor)
+	// Estargz compression.
+	Estargz = internal.NewAlgorithm(types.EstargzChunkedAlgorithmName, types.GzipAlgorithmName, /* Note: InternalUnstableUndocumentedMIMEQuestionMark is not ZstdChunkedAlgorithmName */
```
Comment (mtrmac) [inline pkg/compression/compression.go:41]: The comment is wrong.

(Also, please actually define and document the semantics of that field?)

File: pkg/compression/compression.go [hunk 2]
```diff
@@ -165,3 +171,57 @@ func AutoDecompress(stream io.Reader) (io.ReadCloser, bool, error) {
 	}
 	return res, decompressor != nil, nil
 }
+
+type estarCompressor struct {
```
Comment (mtrmac) [inline pkg/compression/compression.go:175]: Move this into a separate file, please; let’s keep `compression.go` somewhat generic.

File: pkg/compression/compression.go [hunk 3]
```diff
@@ -165,3 +171,57 @@ func AutoDecompress(stream io.Reader) (io.ReadCloser, bool, error) {
...
+	}
+
+	blob, err := estargz.Build(sr, estargz.WithCompressionLevel(compressionLevel))
+	if err != nil {
+		return err
+	}
+	defer blob.Close()
+
+	c.metadata["containerd.io/snapshot/stargz/toc.digest"] = string(blob.TOCDigest())
+	c.metadata["io.containers.estargz.uncompressed-size"] = fmt.Sprintf("%v", st.Size())
```
Comment (mtrmac) [inline pkg/compression/compression.go:206]: Is there a Go package providing these strings as constants?

File: pkg/compression/compression.go [hunk 4]
```diff
@@ -165,3 +171,57 @@ func AutoDecompress(stream io.Reader) (io.ReadCloser, bool, error) {
...
+
+	_, err = io.Copy(c.writer, blob)
+	return err
+}
+
+// estargzCompressor is a CompressorFunc for the estargz compression algorithm.
+func estargzCompressor(r io.Writer, metadata map[string]string, level *int) (io.WriteCloser, error) {
+	file, err := ioutil.TempFile("", "estargz-layer")
```
Comment (mtrmac) [inline pkg/compression/compression.go:214]: `TemporaryDirectoryForBigFiles`

File: pkg/compression/types/types.go [hunk 5]
```diff
@@ -38,4 +38,6 @@ const (
 	// will actually be available. (In fact it is intended for this types package not to depend
 	// on any of the implementations.)
 	ZstdChunkedAlgorithmName = "zstd:chunked"
+	// EstargzChunkedAlgorithmName is the name used by containerd/stargz-snapshotter.
```
Comment (mtrmac) [inline pkg/compression/types/types.go:41]: Add the `NOTE` as well.

**Review by mtrmac — COMMENTED** [review]
How large is the added dependency?

Through some lack of foresight and carelessness, we have ended up with various components that really shouldn’t care, like c/image/docker, depending on this subpackage.

At some point, we need to resolve that, either by a significant reimplementation of those dependency chains, or by making the new compression formats optional.

In this case at least, where we don’t need any new implementation to decompress the data, and where compression using this format can only ever happen by an explicit user request (i.e. `types.SystemContext.CompressionFormat` providing a `*compressiontypes.Algorithm`), it might make sense to move this into a subpackage. And then, to allow simple implementations of CLIs, we might need a `transports/alltransports`-like split of the algorithm registry and the algorithm dependencies.

File: pkg/compression/compression.go (hunk 3 above)
Comment (ktock) [reply pkg/compression/compression.go:206]: Yes, there is at https://github.com/containerd/stargz-snapshotter/blob/v0.7.0/estargz/types.go#L70-L80 .

File: pkg/compression/compression.go (hunk 2 above)
Comment (mtrmac) [inline pkg/compression/compression.go:175]: BTW at least a smoke test that this produces usable results would be nice.

**Comment by giuseppe** [comment]
@ktock do you think we can move some of the logic to github.com/containerd/stargz-snapshotter?

**Comment by ktock** [comment]
@giuseppe I think we can move `estarCompressor` to github.com/containerd/stargz-snapshotter/estargz . Could you open up the PR?

File: pkg/compression/compression.go (hunk 3 above)
Comment (ktock) [inline pkg/compression/compression.go:206]: This needs to be the uncompressed size of the *result* estargz blob (not the source tar).

File: pkg/compression/compression.go [hunk 6]
```diff
@@ -165,3 +171,57 @@ func AutoDecompress(stream io.Reader) (io.ReadCloser, bool, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	// Unlink immediately the file so we won't leak it.
+	os.Remove(file.Name())
```
Comment (AkihiroSuda) [inline pkg/compression/compression.go:219]: Can we return err here?

**Comment by giuseppe** [comment]
I've moved the implementation to `containers/storage`: https://github.com/containers/storage/pull/1001

File: docker/docker_image_src.go [hunk 7]
```diff
@@ -305,53 +305,92 @@ func (s *dockerImageSource) GetBlobAt(ctx context.Context, info types.BlobInfo,
...
-
-	streams := make(chan io.ReadCloser)
-	errs := make(chan error)
+	switch res.StatusCode {
+	case http.StatusOK:
+		// if the server replied with a 200 status code, convert the full body response to a series of
+		// streams as it would have been done with 206.
+		streams := make(chan io.ReadCloser)
+		errs := make(chan error)
+		go func() {
```
Comment (mtrmac) [inline docker/docker_image_src.go:315]: This is becoming rather large (e.g. how far down the “unknown `res.StatusCode` case is); please split the two `streams`-producing goroutines into two separate functions.

Unit tests also look worthwhile, at least for those `streams` producers; without testing actual pulls of this format with high frequency (the way Skopeo and Podman tests trigger at least the default pull paths), we are pretty unlikely to notice breakage.

**Comment by giuseppe** [comment]
@mtrmac moved a patch to a different PR: https://github.com/containers/image/pull/1359

**Pull request closed without merging**